vncviewer/EmulateMB.cxx

```diff
@@ -194,23 +205,198 @@ static const signed char stateTab[11][5][3] = {
+    {  1,  0,  6 },   /* other button -> non emulated, mod down */
+    {  1,  0,  6 }    /* key pressed -> non emulated, mod down */
+  },
+/* 2 button down */
+  {
+    {  1,  0,  5 },   /* mod down -> non emulated, both down */
+    { 10,  0,  0 },   /* mod up -> error, ground */
+    {  0,  0,  0 },   /* left button -> ground */
+    {  0,  0,  2 },   /* other button -> self */
+    {  0,  0,  2 }    /* key pressed -> self */
+  },
+/* 3 emulating both down */
+  {
+    { 10,  0,  0 },   /* mod down -> error, ground */
+    {  0,  0,  4 },   /* mod up -> after emulation, button down, mod up */
+    {  0,  1,  7 },   /* left button -> after emulation, button up, mod down */
+    {  0,  1,  3 },   /* other button -> self */
+    {  0,  0,  3 }    /* key pressed -> self */
+  },
+/* 4 after emulation, button down, mod up */
+  {
+    {  0,  0,  3 },   /* mod down -> emulating both down */
+    { 10,  0,  0 },   /* mod up -> error, ground */
+    {  0,  1,  0 },   /* left button -> error, ground */
+    {  0,  1,  4 },   /* other button -> self */
+    {  0,  0,  4 }    /* key pressed -> self */
+  },
+/* 5 non emulated, both down, mod sent to server */
+  {
+    { 10,  0,  0 },   /* mod down -> error, ground */
+    { -1,  0,  2 },   /* mod up -> button down */
+    {  0,  0,  6 },   /* left button -> non emulated, mod physicall down. button up */
+    {  0,  0,  5 },   /* other button -> self */
+    {  0,  0,  5 }    /* key pressed -> self */
+  },
+/* 6 non emulated, mod physicall down. button up. mod down sent to server */
+  {
+    { 10,  0,  0 },   /* mod down -> error, ground */
+    { -1,  0,  0 },   /* mod up -> ground */
+    { -1,  1,  3 },   /* left button -> emulating both down */
+    {  0,  0,  6 },   /* other button -> self */
+    {  0,  0,  6 }    /* key pressed -> self */
+  },
+/* 7 after emulation, button up, mod down. mod not sent to server */
+  {
+    { 10,  0,  0 },   /* mod down -> error, ground */
+    {  0,  0,  0 },   /* mod up -> ground */
+    {  0,  1,  3 },   /* left button -> emulating both down */
+    {  1,  0,  6 },   /* other button -> non emulated, mod physicall down. button up */
+    {  1,  0,  6 }    /* key pressed -> non emulated, mod physicall down. button up */
+  }
+};
+
+EmulateMB::EmulateMB(rdr::U32 emulateMBModKey)
+  : state(0), emulatedButtonMask(0), timer(this), emulateMiddleButtonModifierKey(emulateMBModKey),
+    mbState(0), mbLastButtonMask(0)
+{
+}
+
+void EmulateMB::mbFsmDoMod(int modAction)
 {
+  if (modAction == 1 || modAction == 2)
+    writeKeyEvent(emulateMiddleButtonModifierKey, modKeyCode, true);
+  if (modAction == -1 || modAction == 2)
+    writeKeyEvent(emulateMiddleButtonModifierKey, modKeyCode, false);
+}
+
+bool EmulateMB::filterKeyPressRelease(bool is_press, int keyCode, rdr::U32 keySym)
+{
+  if (!emulateMiddleButtonMod)
+    return false;
+
+  Action *action;
+  if (keySym == emulateMiddleButtonModifierKey || keyCode == modKeyCode) {
```

Comment:
I would not trust `modKeyCode` unless this is a release. And ideally you know that Alt is pressed.